server/src/services/metadata.service.ts

```diff
@@ -487,6 +496,124 @@ export class MetadataService implements OnEvents {
     }
   }
 
+  async searchPerson(ownerId: AssetEntity['ownerId'], dto: SearchPeopleDto): Promise<PersonResponseDto[]> {
+    return this.personRepository.getByName(ownerId, dto.name, { withHidden: dto.withHidden });
+  }
+
+  private async applyTaggedFaces(asset: AssetEntity, tags: ImmichTags) {
+    this.logger.debug(`Starting faces extraction from tags (${asset.id})`);
+
+    const faceSet = new Set();
+
+    if (!tags.RegionInfo) {
+      return true;
+    }
+
+    interface Region {
+      imageWidth: number;
+      imageHeight: number;
+      boundingBoxX1: number;
+      boundingBoxX2: number;
+      boundingBoxY1: number;
+      boundingBoxY2: number;
+    }
+
+    interface MetadataFaceResult {
+      Name?: string;
+      Type?: string;
+      Region: Region;
+    }
+
+    function toPixels(region: Region, unit: string): Region {
+      const pixelRegion: Region = region;
+      if (unit.toLowerCase() == 'normalized') {
+        pixelRegion.boundingBoxX1 = Math.floor(region.boundingBoxX1 * region.imageWidth);
+        pixelRegion.boundingBoxX2 = Math.floor(region.boundingBoxX2 * region.imageWidth);
+        pixelRegion.boundingBoxY1 = Math.floor(region.boundingBoxY1 * region.imageHeight);
+        pixelRegion.boundingBoxY2 = Math.floor(region.boundingBoxY2 * region.imageHeight);
+      }
+      return pixelRegion;
+    }
+
+    for (const region of tags.RegionInfo?.RegionList) {
+      const faceResult: MetadataFaceResult = {
+        Name: region.Name,
+        Type: region.Type,
+        Region: toPixels(
+          {
+            imageWidth: tags.RegionInfo?.AppliedToDimensions?.W,
+            imageHeight: tags.RegionInfo?.AppliedToDimensions?.H,
+            boundingBoxX1: region.Area.X - region.Area.W / 2,
+            boundingBoxY1: region.Area.Y - region.Area.H / 2,
+            boundingBoxX2: region.Area.X + region.Area.W / 2,
+            boundingBoxY2: region.Area.Y + region.Area.H / 2,
+          },
+          region.Area.Unit,
+        ),
+      };
+
+      faceSet.add(faceResult);
+    }
+
+    const faces = [...faceSet] as MetadataFaceResult[];
+
+    this.logger.debug(`${faces.length} faces detected in ${asset.originalPath}`);
+
+    // cleanup faces imported from metadata so we can update them if we're re-reading to avoid duplications
+    await this.personRepository.deleteFaceFromAsset(asset.id, SourceType.EXIF);
```

Comment:
I dislike that this is outside of a transaction since the job can fail below and still end up deleting these faces. I'd prefer something like an `upsertFaces` method in the repo that encapsulates adding the faces and deleting the old ones in one transaction.

Reply:
Understood, will investigate that approach

Reply:
Done https://github.com/bugfest/immich/commit/c9c85e115d5ebc2d46fcc80d474ae8736b0810d9#diff-bdd10fc89e503edc0aa770142c760f507c6b6817b12b1b3a28df3e7d7145ca69R272-R286